core/commands/dag/dag.go

```diff
@@ -100,52 +97,86 @@ into an object of the specified format.
 			defer n.Blockstore.PinLock().Unlock()
```

Comment:
Given that you add the blocks/pin in a go routine, this needs to go somewhere else (i.e., in the go routine). Also, if you want to improve this slightly, it's probably best not to hold this lock while waiting on the client (either receiving or sending data).